src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp

```diff
@@ -27,34 +28,36 @@ ObjectStorageQueueMetadataFactory::FilesMetadataPtr ObjectStorageQueueMetadataFa
     else
     {
         auto & metadata_from_table = metadata->getTableMetadata();
-        auto & metadata_from_keeper = it->second.metadata->getTableMetadata();
+        auto & metadata_from_keeper = it->second->getTableMetadata();
 
         metadata_from_table.checkEquals(metadata_from_keeper);
-
-        it->second.ref_count += 1;
     }
-    return it->second.metadata;
+
+    it->second->registerIfNot(storage_id);
+    return it->second;
 }
 
-void ObjectStorageQueueMetadataFactory::remove(const std::string & zookeeper_path)
+void ObjectStorageQueueMetadataFactory::remove(const std::string & zookeeper_path, const StorageID & storage_id)
 {
     std::lock_guard lock(mutex);
     auto it = metadata_by_path.find(zookeeper_path);
 
     if (it == metadata_by_path.end())
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Metadata with zookeeper path {} does not exist", zookeeper_path);
 
-    chassert(it->second.ref_count > 0);
-    if (--it->second.ref_count == 0)
+    const size_t registry_size = it->second->unregister(storage_id);
```

Comment:
maybe we should move `registerIfNot` and `unregister` inside the `try catch` block below, as they can fail for keeper fails. what do you think?